### ISSUE TITLE
help with training

### ISSUE THREAD

**Issue by y-x-c** [mindslab-ai/faceshifter#6]
Thanks for the awesome code! I am training my own model right now and have a few questions:

- currently I am using 100k (out of around 1.8m) images from CelebAMask-HQ, ffhq and vggface to train the model. did you use the full set to train your model?
- I didn't see large improvement for most losses anymore (160k steps trained, 4gpus x 12images/batch); is this normal? should I just continue training for more steps?
![image](https://user-images.githubusercontent.com/6591043/98012714-08e33880-1dbf-11eb-9ee2-b0b6ad07deae.png)
![image](https://user-images.githubusercontent.com/6591043/98012739-0ed91980-1dbf-11eb-9dd9-064ffac6b10a.png)
![image](https://user-images.githubusercontent.com/6591043/98012752-139dcd80-1dbf-11eb-977a-9345780275c5.png)
![image](https://user-images.githubusercontent.com/6591043/98012769-17c9eb00-1dbf-11eb-9976-055a4128de68.png)
![image](https://user-images.githubusercontent.com/6591043/98012783-1b5d7200-1dbf-11eb-8c34-c3f3bb337f7f.png)
- I also checked the validation results, and the reconstruction is not good.
![image](https://user-images.githubusercontent.com/6591043/98013578-295fc280-1dc0-11eb-8587-9ec29ed89369.png)
![image](https://user-images.githubusercontent.com/6591043/98013596-2d8be000-1dc0-11eb-89b2-79fc6ffcc086.png)
![image](https://user-images.githubusercontent.com/6591043/98013619-341a5780-1dc0-11eb-85c2-3ba6e696c577.png)
- I noticed `shuffle` for the training dataloader is not set to `True`, did you use the same setting?

Thanks!

**Comment by usingcolor** [comment]
Hi! You did very fast training!

1. Yes, I used full-set dataset. I don't know about IJB-C dataset. The distribution of dataset can influence to your model.

2. In the paper, they trained for 500K steps. I trained for over 500K. In my eye, your losses are getting down for attribute_loss but unstable for Rec and ID loss. In my case, the two losses are more stable and lower at the same steps.

3. `shuffle` option in training dataloader should be `True`. It is clearly my mistake while publishing.

**Comment by y-x-c** [comment]
> Hi! You did very fast training!
> 
> 1. Yes, I used full-set dataset. I don't know about IJB-C dataset. The distribution of dataset can influence to your model.
> 2. In the paper, they trained for 500K steps. I trained for over 500K. In my eye, your losses are getting down for attribute_loss but unstable for Rec and ID loss. In my case, the two losses are more stable and lower at the same steps.
> 3. `shuffle` option in training dataloader should be `True`. It is clearly my mistake while publishing.

Thanks for your reply.
1. I just corrected the description, I am using the same datasets (CelebAMask-HQ, ffhq and vggface) as well.

2. 
  - So in your case, each step has 64 images; and let's say there are 1.5m images in those three datasets, so you trained for around 4 epochs (= 64 * 500000 / 1500000 / 5 ) in total? 
  -  In my case, each step only has 48 images, so maybe that's why the two losses are higher at the same steps.
  - I found the Rec loss is going much lower in the third epoch, and the results are much better than before. I will continue my current training and see what's going on.
3. Thanks for the clarification, I also changed to `True` during my training.

**Comment by usingcolor** [comment]
- I trained with 32 batch size, it is the same as the paper. (Two V100 32G GPUs, 16 batch size for each)

- Training GAN is very unstable. If your loss is going down, I think it works well.

**Comment by Qiulin-W** [comment]
> > Hi! You did very fast training!
> > 
> > 1. Yes, I used full-set dataset. I don't know about IJB-C dataset. The distribution of dataset can influence to your model.
> > 2. In the paper, they trained for 500K steps. I trained for over 500K. In my eye, your losses are getting down for attribute_loss but unstable for Rec and ID loss. In my case, the two losses are more stable and lower at the same steps.
> > 3. `shuffle` option in training dataloader should be `True`. It is clearly my mistake while publishing.
> 
> Thanks for your reply.
> 
> 1. I just corrected the description, I am using the same datasets (CelebAMask-HQ, ffhq and vggface) as well.
> 
> * So in your case, each step has 64 images; and let's say there are 1.5m images in those three datasets, so you trained for around 4 epochs (= 64 * 500000 / 1500000 / 5 ) in total?
> * In my case, each step only has 48 images, so maybe that's why the two losses are higher at the same steps.
> * I found the Rec loss is going much lower in the third epoch, and the results are much better than before. I will continue my current training and see what's going on.
> 
> 1. Thanks for the clarification, I also changed to `True` during my training.

Hi, did you change the coefficients of different loss terms? I found my training unstable with the coeffs provided by the author...

**Comment by hanikh** [comment]
> * I trained with 32 batch size, it is the same as the paper. (Two V100 32G GPUs, 16 batch size for each)
> * Training GAN is very unstable. If your loss is going down, I think it works well.

it means that you have used 'dp' instead of 'ddp'?
since in 'ddp' mode the whole batch is not devided between GPUs.

**Comment by payne4handsome** [comment]
@y-x-c Hi, have you got the satisfying result? I trained just with FFHQ and CelebA-HQ datasets about 90 thousand images. The result is bad just like below.
![image](https://user-images.githubusercontent.com/28085731/103514070-d8934580-4ea6-11eb-866c-1f7d68a9d6f2.png)
![image](https://user-images.githubusercontent.com/28085731/103514087-dfba5380-4ea6-11eb-9db3-d266d235be3e.png)
![image](https://user-images.githubusercontent.com/28085731/103514130-fcef2200-4ea6-11eb-8b92-b9599bd8addc.png)

**Comment by princessmittens** [comment]
By 4 epoch's you mean 26...

I am about two weeks into training at about the halfway mark. I noticed that some of the results on the colab show some image artifacts. Is that present in all final results? Did you manage to fix that with more training?

**Comment by hanikh** [comment]
> By 4 epoch's you mean 26...
> 
> I am about two weeks into training at about the halfway mark. I noticed that some of the results on the colab show some image artifacts. Is that present in all final results? Did you manage to fix that with more training?

@princessmittens I am  also working on this paper and I have some questions about this implementation. May I have your email address?

**Comment by princessmittens** [comment]
@usingcolor I am near the end of training and these are my results. I have trained on 8 gpu's with 32 gig ram and a 21 batch size/ per gpu. The results have been pretty bad so far. I tried my best to recreate the exact parameters with all 3 datasets (~1.3 million images after processing) and have trained for about 2-3 weeks. With my current batch size and according to the results, I'm at the 79% mark in reference to 500k. 

@y-x-c -Have you been able to recreate better results?  Is it worth continuing? 

This has cost a lot of money/time Any input would be great. 

@hanikh Not sure how much I can help you considering my results but my email is <>

Src:
![source](https://user-images.githubusercontent.com/35903814/110513605-cc14ae80-80dc-11eb-9285-50d3b83fff47.png)
Target:
![target](https://user-images.githubusercontent.com/35903814/110513763-f6ff0280-80dc-11eb-93e1-08d3c572542d.png)
![target2](https://user-images.githubusercontent.com/35903814/110513774-f9615c80-80dc-11eb-829d-93830241d126.png)
Results: 
![testimg2](https://user-images.githubusercontent.com/35903814/110513531-b4d5c100-80dc-11eb-86a9-1a820f0b0c26.png)
![testimg](https://user-images.githubusercontent.com/35903814/110513548-bacba200-80dc-11eb-9c8f-c51de7c42ffd.png)

**Comment by tamnguyenvan** [comment]
Hello, anyone here got good result?

**Comment by princessmittens** [comment]
No-I have talked to @hanikh. I don't think anyone has been able to recreate the results as of yet.

**Comment by ghost** [comment]
I have better results than this, princessmittens, can you leave your e-mail and I will reach out to you?

> No-I have talked to @hanikh. I don't think anyone has been able to recreate the results as of yet.

**Comment by tamnguyenvan** [comment]
@cwalt2014 Can you please share the source code or pretrained weights? I will appreciate that. My email: tamvannguyen200795@gmail.com
Thanks.

**Comment by DeliaJIAMIN** [comment]
@cwalt2014  Dear friend！Can you please share the source code? Thanks a million🙏🙏. My email: zhangjiajia827@gmail.com
Thanks！🙏🙏🙏

**Comment by princessmittens** [comment]
@cwalt2014 my email is andreachristians@gmail.com

**Comment by ZhiluDing** [comment]
@cwalt2014 my email is dzl0418@gmail.com

**Comment by ghost** [comment]
@cwalt2014 I would also love to know what changes you would suggest to get better results 🙏 my mail is paulchvn@gmail.com

**Comment by Seanseattle** [comment]
@cwalt2014 Could you please share the source code or pretrained weights? Thank you. My Email: seanzlxu@gmail.com

**Comment by chinasilva** [comment]
@cwalt2014 Thank you very much. My Email:476369545@qq.com

**Comment by lefsiva** [comment]
@cwalt2014 Thank you. My email is: lefsiva7@gmail.com

**Comment by akafen** [comment]
@cwalt2014 Thank you very much. My email is:  302926535@qq.com

**Comment by tyrink** [comment]
> I have better results than this, princessmittens, can you leave your e-mail and I will reach out to you?
> 
> > No-I have talked to @hanikh. I don't think anyone has been able to recreate the results as of yet.

Hi, @cwalt2014, could you please send me some of your results? I wonder what the possible results looks like. My email is 1085425753@qq.com. Any reply will be appreciated.

**Comment by suzie26** [comment]
@cwalt2014 Can you share your code or pretrained weights?? Thank you soooo much!!
My email is: suzieya26@gmail.com

**Comment by Daisy-Zhang** [comment]
@cwalt2014 Could you share your code or pretrained weights?? Thank you very much!! My email is: daisy.zdcc@gmail.com

**Comment by usingcolor** [comment]
@Daisy-Zhang @suzie26 @tyrink @akafen @lefsiva @chinasilva @Seanseattle @Poloangelo @ZhiluDing @princessmittens @DeliaJIAMIN @tamnguyenvan @cwalt2014  Check out  [HifiFace](https://github.com/mindslab-ai/hififace), our implementation of a more recent face-swapping model with the pre-trained model.

**Comment by chuer-yu** [comment]
@cwalt2014 hello, could you please share your pretrained weights? Thank you so much! My email is: chuer.yu1995@gmail.com

**Comment by ywon0925** [comment]
@cwalt2014 could you please share your pre-trained weights? I would really appreciate it! My email is: 9788667@gmail.com

**Comment by niuyuanc** [comment]
@antonsanchez  Could you please share the pretrained weights? Thank you so much 🙏🙏🙏 
My email: niuyuanc@163.com
Thanks🙏🙏🙏

**Comment by galmizush** [comment]
@cwalt2014 could you please share your pre-trained weights? I would really appreciate it! My email is: [galmizush@gmail.com](mailto:galmizush@gmail.com)

**Comment by Jaep0805** [comment]
@cwalt2014 could you please share your pre-trained weights? My email is: jaep0805@snu.ac.kr
Thank you so much

**Comment by dsj320** [comment]
> I have better results than this, princessmittens, can you leave your e-mail and I will reach out to you?我的结果比这个更好，princessmittens，你能留下你的电子邮件吗？我会联系你的。
> 
> > No-I have talked to [@hanikh](https://github.com/hanikh). I don't think anyone has been able to recreate the results as of yet.我已经和某人谈过了。我不认为到目前为止还有谁能重现这些结果。
could you please share your pre-trained weights? My email is: dushuangjun2024@ia.ac.cn,Thank you very much!
